api/config/config.go

```diff
@@ -16,36 +16,42 @@ package config
 
 import (
 	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
 	"time"
 
-	"github.com/kelseyhightower/envconfig"
-	v1 "k8s.io/api/core/v1"
-	resourcev1 "k8s.io/apimachinery/pkg/api/resource"
-
 	"github.com/caraml-dev/merlin/pkg/transformer/feast"
 	"github.com/caraml-dev/merlin/pkg/transformer/spec"
 	internalValidator "github.com/caraml-dev/merlin/pkg/validator"
 	mlpcluster "github.com/caraml-dev/mlp/api/pkg/cluster"
 	"github.com/caraml-dev/mlp/api/pkg/instrumentation/newrelic"
 	"github.com/caraml-dev/mlp/api/pkg/instrumentation/sentry"
+	"github.com/go-playground/validator"
+	"github.com/mitchellh/mapstructure"
+	"github.com/ory/viper"
+	"gopkg.in/yaml.v2"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	resourcev1 "k8s.io/apimachinery/pkg/api/resource"
 )
 
 const (
 	MaxDeployedVersion = 2
 )
 
 type Config struct {
-	Environment           string          `envconfig:"ENVIRONMENT" default:"dev"`
-	Port                  int             `envconfig:"PORT" default:"8080"`
-	LoggerDestinationURL  string          `envconfig:"LOGGER_DESTINATION_URL"`
-	Sentry                sentry.Config   `envconfig:"SENTRY" split_words:"false"`
-	NewRelic              newrelic.Config `envconfig:"NEWRELIC" split_words:"false" `
-	EnvironmentConfigPath string          `envconfig:"DEPLOYMENT_CONFIG_PATH" required:"true"`
-	NumOfQueueWorkers     int             `envconfig:"NUM_OF_WORKERS" default:"2"`
-	SwaggerPath           string          `envconfig:"SWAGGER_PATH" default:"./swagger.yaml"`
-
-	DeploymentLabelPrefix string `envconfig:"DEPLOYMENT_LABEL_PREFIX" default:"gojek.com/"`
-	PyfuncGRPCOptions     string `envconfig:"PYFUNC_GRPC_OPTIONS" default:"{}"`
+	Environment           string `default:"dev"`
```

Comment:
Okay I've added that `reflectViperConfig` method though it needed a tiny change to not call `v.SetDefault` under the hood for **_all_** fields.  In particular, the `mlp.K8sConfig` field/struct (of the image builder config) gets initialised with zero-values when `v.SetDefault` gets called on it (it stays as such if it doesn't get populated by non-zero values in the other config files). This causes a problem later when the struct gets converted to an interface, and then marshalled into a yaml byte string, which causes a panic with the following error:
```
panic: reflect.Value.Interface: cannot return value obtained from unexported field or method
```
Previously, when `reflectViperConfig` wasn't used, this wasn't an issue because the `v.SetDefault` wouldn't be creating a zero-valued `mlp.K8sConfig` field that would be parsed by the marshaller. 

A quick fix to this was to create a unique `default:"-"` tag-value pair that simply causes `reflectViperConfig` to ignore that particular field and not call `v.SetDefault` on it. This is applied to the `mlp.K8sConfig` field.

Reply:
TIL. I agree that the first method is better.

Btw, whey are we setting `default:"-"` for some other fields too, like `FeastRedisConfig`, `FeastBigtableConfig`, etc. ?

Reply:
Because (I believe) these fields are meant to be optional (as seen from some of the unit tests), and if we don't skip calling `v.SetDefault` on them, those fields will have their nested fields initialised with zero values, like `FeastRedisConfig.ServingURL` (a string), and their own `validate:"required"` tags will invalidate themselves because the tag forces their values to be non-zero.

Reply:
Ahh got it. Makes sense. Thanks for the clarification.